k2/csrc/fsa_algo.cc

```diff
@@ -540,4 +542,85 @@ bool TopSort(const Fsa &a, Fsa *b,
   return true;
 }
 
+std::unique_ptr<Fsa> CreateFsa(const std::vector<Arc> &arcs) {
+  if (arcs.empty()) return nullptr;
+
+  std::vector<std::vector<Arc>> vec;
+  for (const auto &arc : arcs) {
+    auto src_state = arc.src_state;
+    auto dest_state = arc.dest_state;
+    auto new_size = std::max(src_state, dest_state);
+    if (new_size >= vec.size()) vec.resize(new_size + 1);
+    vec[src_state].push_back(arc);
+  }
+
+  std::stack<DfsState> stack;
+  std::vector<char> state_status(vec.size(), kNotVisited);
+  std::vector<int32_t> order;
+
+  auto num_states = static_cast<int32_t>(vec.size());
```

Comment:
it's not necessary to covert type here, keep all things with `size_t` is better as when you read `state_status[i]`, it will be not converted back to `size_t` again.

Reply:
- `static_cast` does not incur runtime cost in this case.
- I need to use `auto i = 0;` in subsequent `for` loops.

Reply:
but `i=0` is OK for size_t.

Reply:
I tried to remove `static_cast<int32_t>` and change `auto i = 0` to `size_t i = 0`.

But the problem**s** are that

- `stack.push`, I have to use `static_cast<int32_t>(i)`
- For `std::vector<int32_t> old_to_new`, I have to use old_to_new[order[i]] = static_cast<int32_t>(i)
- If I use `std::vector<size_t>` for `old_to_new`, I have to use `static_cast<int32_t>` for
arc.src_state and arc.dest_state

-----

You should avoid using unsigned integer types to your best effort.


From Google C++ code style https://google.github.io/styleguide/cppguide.html#Integer_Types

> You should not use the unsigned integer types such as uint32_t, unless there is a valid reason such as representing a bit pattern rather than a number, or you need defined overflow modulo 2^N. In particular, do not use unsigned types to say a number will never be negative. Instead, use assertions for this.

Reply:
Oh, I see, you will `i` in `DfsState` as `int32_t`

Reply:
Some languages, such as Java, do NOT have unsinged integer types.

Reply:
I agree with this guide except size_t as all of STL containers' size_type is size_t, using `size_t` if possible will make code very clean.

Of course, you have to use `int32_t` as it will be pushed, sorry.

Reply:
BTW, for Java, that's why people added `parseUnsignedInt`, `parseUnsignedLong` to JDK. Also many people want an unsigned byte in Java because they would comunicate with network or C.